### ISSUE TITLE
Middleware are not executed on Starlite error

### ISSUE THREAD

**Issue by vrslev** [starlite-api/starlite#175]
When route is not found, method is not allowed, or some similar error, then middleware are not executed although and it is unexpected behaviour.

Starlite, second test fails:
```python
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response
from starlite import Starlite, TestClient, get


class ServerHeaderMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: RequestResponseEndpoint
    ) -> Response:
        response = await call_next(request)
        response.headers["Server"] = "NULL"
        return response


@get()
async def route() -> str:
    return ""


app = Starlite(
    debug=True,
    route_handlers=[route],
    middleware=[ServerHeaderMiddleware],
)


def test_middleware_is_executed_when_route_found():
    with TestClient(app) as client:
        response = client.get("/")
        assert response.headers["Server"] == "NULL"


def test_middleware_is_executed_when_route_not_found():
    with TestClient(app) as client:
        response = client.get("/non-existing-route")
        assert response.headers["Server"] == "NULL"
```

Starlette, tests pass:
```python
from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import PlainTextResponse, Response
from starlette.routing import Route
from starlette.testclient import TestClient


class ServerHeaderMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: RequestResponseEndpoint
    ) -> Response:
        response = await call_next(request)
        response.headers["Server"] = "NULL"
        return response


async def route(request: Request):
    return PlainTextResponse("")


app = Starlette(
    debug=True,
    routes=[Route("/", route, methods=["GET"])],
    middleware=[Middleware(ServerHeaderMiddleware)],
)


def test_middleware_is_executed_when_route_found():
    with TestClient(app) as client:
        response = client.get("/")
        assert response.headers["Server"] == "NULL"


def test_middleware_is_executed_when_route_not_found():
    with TestClient(app) as client:
        response = client.get("/non-existing-route")
        assert response.headers["Server"] == "NULL"
```

**Comment by vrslev** [comment]
Because of this problem CORS headers also are applied to response. This leads to problems in requests from browser.

**Comment by vrslev** [comment]
Starlite handles raised exceptions here:
https://github.com/starlite-api/starlite/blob/935452796cf228b9f19f7cbd70ee45b540da00f7/starlite/app.py#L150-L151

The problem is that it skips middleware stack.
Starlette on other hand manages exception handlers with `ExceptionMiddleware`:
https://github.com/encode/starlette/blob/5b058faa108ebe0e7d8810e84bf88cb37744faf8/starlette/middleware/exceptions.py#L11
Middleware stack also looks differently:
https://github.com/encode/starlette/blob/3453fd6b6259393f3744815b49af297c934f32d0/starlette/applications.py#L91-L99

**Comment by vrslev** [comment]
@Goldziher I like the idea of having `ExceptionMiddleware`. Should we move `Starlite.handle_exception` logic to separate middleware and build stack with it on top?

**Comment by peterschutt** [comment]
This is good you bring this up now, cheers. There is some talk about middleware stack refactoring related to #169 and this will be good to think about at the same time.

Have you seen #173?

**Comment by Goldziher** [comment]
> @Goldziher I like the idea of having `ExceptionMiddleware`. Should we move `Starlite.handle_exception` logic to separate middleware and build stack with it on top?

No - we allow exception handling on multiple layers. This cannot live inside middleware. Also its less elastic to do so. We will need a different solution

**Comment by vrslev** [comment]
Just looked at #173 — that is a great idea! It won’t affect performance much because middleware stacks are built on initialization, right?

**Comment by vrslev** [comment]
If we want middleware to run on errors, we can't catch them in middleware.

**Comment by Goldziher** [comment]
@vrslev please checkout --> https://github.com/starlite-api/starlite/pull/173, I add a fix there.

**Comment by vrslev** [comment]
Sure. Did you already added a fix or will work on that in the PR? I can't tell because you force-pushed 😅

**Comment by Goldziher** [comment]
> Sure. Did you already added a fix or will work on that in the PR? I can't tell because you force-pushed 😅

I added a fix, look in app.py

**Comment by vrslev** [comment]
I think the fix you suggested breaks abstraction layers. It only covers CORS and wouldn't work with the example I suggested in this issue.
https://github.com/starlite-api/starlite/blob/66684947c43c84c107699ce84799a038f7244442/starlite/app.py#L161-L162

**Comment by Goldziher** [comment]
Then what do you suggest?

**Comment by vrslev** [comment]
What about layered `ExceptionMiddleware`? Add it for each layer where exception handlers are accepted

**Comment by vrslev** [comment]
First, I think we should determine whether errors in middleware should be handled by exception handlers. If it is so, then it is impossible to modify response on internal error.

**Comment by Goldziher** [comment]
ok, i left this out of my PR for now, we will deal with it separately once there is a clear plan on how to do this.

**Comment by vrslev** [comment]
@Goldziher Do you think exception handlers should handle errors in middleware?

**Comment by Goldziher** [comment]
I think they should receive the exceptions and handle them no matter the origin. The reason is that exception handlers need to handle the exception -> response logic.

But we still need a solution to this issue-

That's why I suggest to move CORS from a middleware to be a part of the app layer in starlite.

**Comment by vrslev** [comment]
It is kinda ugly, but what about some kind of `final_middleware` that run even on exceptions?
By the way, middleware run even on exceptions in Starlette (https://www.starlette.io/middleware/).

**Comment by vrslev** [comment]
I think we shouldn't make CORS a special middleware that runs even after exception handlers. There're plenty use cases where similar functionality would be required.

**Comment by Goldziher** [comment]
> It is kinda ugly, but what about some kind of `final_middleware` that run even on exceptions? By the way, middleware run even on exceptions in Starlette (https://www.starlette.io/middleware/).

Thats not precise - in starlette there is an exceptions middleware that captures exceptions from the application and transforms them into responses. You can see this logic here: https://github.com/encode/starlette/blob/master/starlette/applications.py#L78. This though limits our ability to capture exceptions raised in middleware.

We could inject the exception handler resolution logic as the final middleware layer, as is done in Starlette, but we will be limiting ourselves to the kind of logic that can be ran here.

**Comment by vrslev** [comment]
I think loosing exception handling in middleware to make middleware run no matter what (which is expected behaviour from my perspective) is an expedient tradeoff.

**Comment by peterschutt** [comment]
I've built a POC of Exception middleware based on #173.

I think it will help having something to discuss.

https://github.com/starlite-api/starlite/compare/issue-169-add-handler-middleware-support...exception-middleware

**Comment by Goldziher** [comment]
So, as I commented in the PR linked above - this doesn't factor in the way exceptions are handled in the `route`. Namely, 
 the code from `starlite/route.py#L147`, which includes the route handler per route. This is only for http routes mind you - although we could consider expanding this to the other routes as well.

```python
    async def call_handler(
        self, scope: Scope, request: Request, parameter_model: KwargsModel, route_handler: HTTPRouteHandler
    ) -> Union[Response, StarletteResponse]:
        """
        Calls the before request handlers, retrieves any data required for the route handler,
        and calls the route handler's to_response method.

        This is wrapped in a try except block - and if an exception is raised,
        it tries to pass it to an appropriate exception handler - if defined.
        """
        try:
            response_data = None
            before_request_handler = route_handler.resolve_before_request()
            # run the before_request hook handler
            if before_request_handler:
                if is_async_callable(before_request_handler):
                    response_data = await before_request_handler(request)
                else:
                    response_data = await run_sync(before_request_handler, request)
            if not response_data:
                response_data = await self.get_response_data(
                    route_handler=route_handler, parameter_model=parameter_model, request=request
                )
            return await route_handler.to_response(
                app=scope["app"],
                data=response_data,
                plugins=request.app.plugins,
            )
        except Exception as e:
            handler = get_exception_handler(route_handler.resolve_exception_handlers(), e)
            if handler:
                return handler(request, e)
            raise e
```

The question really boils down to this - do we need to have middlewares running on exception responses as well? 

Lets set CORS aside, because there we know the answer is yes. For my money I'm not sure we do - I would simply remove CORS from being considered a middleware, to run after either a reponse or error response are generated. 

We could consideer adding more lifecycle hooks, in this case - an `afterError` hook to allow processing logic specifically for errors, so if users want to modify the error responses in some ways without using an exception handler to do this, they could potentially do it form there (or do other stuff, like prometheus integrations etc.). 

Thoughts?

**Comment by peterschutt** [comment]
> By the way, I really slapped this together with very little thought, so I'm sure there are problems but gives us somewhere to start.

Yeh, so that branch was never intended to be a PR, I'm just trying to understand the issue and talking about real code just helps to move things out of the abstract.

So, I'll keep it high level for now so we don't get distracted.

> The question really boils down to this - do we need to have middlewares running on exception responses as well?

Right now, my gut says yes. For starters the docs say:

 > Middlewares are useful when you need to operate on **all incoming requests** on the app level.

Further to that, it is a pretty well understood paradigm - I think it is what users will mostly expect when they come to us from other solutions in the ecosystem.

> Thoughts?

I'm just still trying to understand the overall problem space before I'll commit to solutions. At the highest level, I see two options, feel free to add/remove or debate on the below, but as I see it we:

1. Do more in the context of the middleware stack - particularly route resolution so that `404`'s and `405`s are raised within that context and so any middlewares that should run before the route is resolved would be executed.
2. Do more outside the context of the middleware stack with assorted callbacks and handlers for handing particular cases.

With #173 there will be a subset of middleware that we are aware of before the routing has occurred (that is the middleware registered on the app itself), and a subset only known after the handler is resolved. In either case above, running the middleware registered on the app before path resolution occurs make sense.

What about where we have middleware registered on the app, and on a controller. We resolve the path to the controller, but then there is no handler for the request method. Should the middleware for the controller be executed in that case? It is pretty much the same problem as the application level middleware running before anything else, but just a few layers deeper in the hierarchy.

**Comment by peterschutt** [comment]
Sorry if a bit terse but I stopped mid-workout to get the idea out before I lose it, but just an idea for our conceptual understanding of what middleware is, and how we treat it.

This comes down to an interpretation of middleware, is it after the server, but before any application interaction whatsoever, or can it be executed between routing/validation and handling (i.e is execution subject to client error) and still be called middleware? They are two different things and we are trying to support both.

What if we say this. 

If you register the middleware on the controller:
- we treat it the same as starlette middleware - exception handling wraps the middleware stack and the middleware is guaranteed to run unless the middleware isn't otherwise reached.
- no routing or validation occurs before these middleware's are called.
- if something in this stack returns, or raises, the Starlite app isn't even reached - so we can't support any niceties like pre-validating and parsing parameter values.

IF middleware registered at or below the router level:
- existing error handling technique
- runs only if no client error, 4xx
- we support pre-validation and parsing of parameter values and they are made available per resolving #189

Starlette middleware would only be supported at the app level.

Our style middleware can be either at the app level, but also specified all the way down to the handler - but no guarantee of execution.

We handle CORS and whatever else the best way we can within the above context, I.e, starlette style middleware, or our style middleware or as an app layer construct.

**Comment by peterschutt** [comment]
We could call it `RoutedMiddleware` and have it be the canonical way to wrap the context of a specific handler's execution, not the entire App's execution.

**Comment by MrAwesomeRocks** [comment]
I just read through this, and while the two different middleware types seem like a good idea, it still leaves a problem: how would you add a middleware that does something like `response.headers["Server"] = "NULL"` on error responses?

Things like the header editor and CORS aren't "advanced" middlewares though; they don't need routing or validation. But they need to run on every request, so they are "global" middlewares. So it seems that all that is needed is a third way to add a "basic" middleware to the entire app.

---

Maybe we make two types of middleware:
1. "Basic" middleware that needs to run every time, but doesn't need routing
2. "Advanced" middleware that gets validation/parameters/etc.

And then have `add_basic_middleware` and `add_advanced_middleware` everywhere where there was `add_middleware`, not limiting basic middleware to just the controller or advanced middleware to everywhere else.

To avoid a breaking change, add a `advanced=True` parameter to `add_middleware` and then just delegate to the other two functions.

**Comment by peterschutt** [comment]
> how would you add a middleware that does something like response.headers["Server"] = "NULL" 

Explain to me why you couldn't?

**Comment by peterschutt** [comment]
> Maybe we make two types of middleware:

I'll push back on the complexity. Everyone knows how starlette's middleware works - no-one knows what Basic and Advanced middleware is.

**Comment by peterschutt** [comment]
> how would you add a middleware that does something like response.headers["Server"] = "NULL"

Also this could be done via uvicorn: https://stackoverflow.com/a/65263873/6560549

**Comment by vrslev** [comment]
> What if we say this.
> 
> If you register the middleware on the controller:
> 
> we treat it the same as starlette middleware - exception handling wraps the middleware stack and the middleware is guaranteed to run unless the middleware isn't otherwise reached.
> no routing or validation occurs before these middleware's are called.
> if something in this stack returns, or raises, the Starlite app isn't even reached - so we can't support any niceties like pre-validating and parsing parameter values.
> IF middleware registered at or below the router level:
> 
> existing error handling technique
> runs only if no client error, 4xx
> we support pre-validation and parsing of parameter values and they are made available per resolving https://github.com/starlite-api/starlite/issues/189
> Starlette middleware would only be supported at the app level.

It makes sense. So, there would be app-wide middleware _and_ other (router, controller, route)?

**Comment by vrslev** [comment]
> Also this could be done via uvicorn: https://stackoverflow.com/a/65263873/6560549

“Server” header is just an example. Currently we can’t use https://github.com/bigbag/starlette-request-id or https://github.com/cak/secure. I’m not even sure if Sentry ASGI integration works

**Comment by peterschutt** [comment]
Sentry works - I use it already, just pass to middleware on the app 👍 there's a typing issue with it though in pycharm but no biggy.

> “Server” header is just an example.

Well I'm just playing the hand that's in-front of me - I probably won't spend any more time on thinking about this for the rest of the day. Feel free to detail the issues with those two handlers and some potential solutions.

**Comment by peterschutt** [comment]
> So, there would be app-wide middleware and other (router, controller, route)?

![image](https://user-images.githubusercontent.com/20659309/176348375-98ed266b-4c83-4293-88c5-20c28a86b895.jpeg)

Sorry about the image :laughing: That's pretty much what I'm thinking. It would be nice if the response could find its way back to the traditional middleware stack after exiting "innerware" stack so it could be inspected/modified but I'd have to have a look to know if possible.. could that be done?

**Comment by Goldziher** [comment]
Couldn't quite get your white boarding. 

Anyhow, @peterschutt your idea is basically to have top level Middleware that is executed after exception handlers, and lower level middlewares that are executed before, correct? 

The issue I have with this is that it's less intuitive and will be a cause of headaches for both users and maintainers. It would be better to avoid IMO. 


The Express Middleware pattern, which is basically what Starlette has, is very simple. We basically departed from it the moment we went for layers Middleware. 

What I suggest instead is to inject exception Handling before all the middlewares, and do this at the asgi router level. 

Basically we will mirror what Starlette is doing - there will be a sandwich of two exception handlers - exception handlers that wrap the route handler and it's dependencies, and an exception handler that wraps the middleware. 

We can define the either separately, or keep it simple and reuse the exception handlers.

**Comment by vrslev** [comment]
> What I suggest instead is to inject exception Handling before all the middlewares, and do this at the asgi router level.

Would middleware run on exception then?

**Comment by peterschutt** [comment]
> Couldn't quite get your white boarding.

Feel free to ask me to explain anything.

> The issue I have with this is that it's less intuitive

We say this  in the docs:

> Starlite builds on top of the [Starlette middleware architecture](https://www.starlette.io/middleware/) and is 100% compatible with it - and any 3rd party middlewares created for it.

Yet we get issues like this one and comments like this:

> Currently we can’t use https://github.com/bigbag/starlette-request-id or https://github.com/cak/secure. I’m not even sure if Sentry ASGI integration works

...that say otherwise.

So, I'm trying to demonstrate that we are trying to do two different things and call them the same thing - and make it work. Middleware that is independent of routing (100% compatible with starlette), and middleware that is totally dependent on it (404/5's don't trigger middleware).

Recognizing the difference also gives us a neat way to work around the compatibility issues with providing a typed set of parameters and an interface to the parsed and validated values as you mentioned in discord and is the subject of #189 - that being they would be separate protocols.

> What I suggest instead is to inject exception Handling before all the middlewares, and do this at the asgi router level.

No problems, I feel like I've demonstrated my point well enough that I'm happy to let go of it if you aren't on board, so If that's the call then lets do it. If you want to do up a checklist of the work involved I can pitch in.

**Comment by Goldziher** [comment]
Why are we incompatible with these middlewares? I don't know that we aren't. Sentry works for sure, why isn't the other one working?

**Comment by peterschutt** [comment]
Yeh I'm using sentry too, it was just a copy-pasta.

I don't know, and I don't really have the time available to find out - if @vrslev could help explain that'd be good.

But doesn't the mere fact that we swallow exceptions rather than propagate them mean that we aren't 100% compatible with starlette middleware? We support their protocol/interface, but the behavior is different, no?

**Comment by Goldziher** [comment]
> Yeh I'm using sentry too, it was just a copy-pasta - and i did defend that earlier if you want to see that.
> 
> > why isn't the other one working?
> 
> I don't know, and I don't really have the time available to find out - if @vrslev could help explain that'd be good.
> 
> But doesn't the mere fact that we swallow exceptions rather than propagate them mean that we aren't 100% compatible with starlette middleware? We support their protocol/interface, but the behavior is different, no?

We have a bug for sure. We are nonetheless compatible with their Middleware, which is what we claim. We do not claim to be Starlette, nor are we going to be Starlette. 

Look, I'm trying to have a pragmatic and factual discussion, and reach for a good architecture, nothing more and nothing less. All we need are open minds and a capacity for discussion.

**Comment by peterschutt** [comment]
> Look, I feel you are defensive.

Not at all, you asked for my opinion and I've given it. I'm trying to conceptualize the problem and all my points and questions are purely aimed toward that. So far, you've not directly addressed any point I've made but have made it clear that you don't agree, and that's fine. Did you see this:

> No problems, I feel like I've demonstrated my point well enough that I'm happy to let go of it if you aren't on board, so If that's the call then lets do it. If you want to do up a checklist of the work involved I can pitch in.

**Comment by vrslev** [comment]
> I don't know, and I don't really have the time available to find out - if @vrslev could help explain that'd be good.

Sorry, I just assumed Sentry might not be working. I didn’t have a chance to test it yet.

**Comment by Goldziher** [comment]
Well, I think what you're proposing is viable, but it places some conceptual complexities for the user. From my PoV it's better to abstract complexity into our end. Saying that, it needs to be discussed.

**Comment by Goldziher** [comment]
> > I don't know, and I don't really have the time available to find out - if @vrslev could help explain that'd be good.
> 
> Sorry, I just assumed Sentry might not be working. I didn’t have a chance to test it yet. 

You mentioned another Middleware not working?

**Comment by peterschutt** [comment]
> Well, I think what you're proposing is viable, but it places some conceptual complexities for the user. From my PoV it's better to abstract complexity into our end. Saying that, it needs to be discussed.

Sure, lets get started with rolling out the solution as you see it. It will help me to understand your proposal better and we can discuss as it evolves.

**Comment by vrslev** [comment]
> You mentioned another Middleware not working?

Yes. https://github.com/bigbag/starlette-request-id, https://github.com/cak/secure — they both won’t work on exceptions.

**Comment by Goldziher** [comment]
> > You mentioned another Middleware not working?
> 
> Yes. https://github.com/bigbag/starlette-request-id, https://github.com/cak/secure — they both won’t work on exceptions.

Ok, but that's the known bug.

**Comment by Goldziher** [comment]
PR added, please review --> https://github.com/starlite-api/starlite/pull/197

**Comment by vrslev** [comment]
@Goldziher I will do it tonight or tomorrow, thanks!

**Comment by Goldziher** [comment]
Ok, we discussed it and for now we will wait to see if this is indeed an issue. CORS will run on these errors - because that happens outside of the middleware stack. We might add a lifecycle hook in the future that allows modifying these in the ASGI router. For now though we will keep behaviour as is. 

Thanks for testing!

**Comment by vrslev** [comment]
@Goldziher Then I don’t understand the point of changes in #197. Nor my example nor something like https://github.com/bigbag/starlette-request-id would work as expected 🤔

**Comment by Goldziher** [comment]
> @Goldziher Then I don’t understand the point of changes in #197. Nor my example nor something like https://github.com/bigbag/starlette-request-id would work as expected 🤔

They will work fine on anything that is not a 404/405 raised by the asgi router. 

Why would you need to set these headers on these type of errors?

**Comment by vrslev** [comment]
I want to use [_secure_ package](https://github.com/cak/secure) on all incoming requests, not only when route is found.

**Comment by Goldziher** [comment]
I dont think these headers specifically are important for these errors, but to each his own. You can set the headers in an exception handler as stated above. Its true that you cannot do this using middleware, but we cannot eat the cake and have it whole.

**Comment by vrslev** [comment]
> If you think it's important to set these headers on 404/5 errors returned by the ASGI router, you can create an app level exception handler to handle these.

Sure there are workarounds. I'm more worried about other users that would be surprised by the current behaviour.

**Comment by Goldziher** [comment]
Yes its not perfect. But there is no way really to make it perfect - its either we restrict middleware in the way it is restricted in Starlette, add a different layer of middleware that is applied to these errors, apply the app layer middleware to these errors, or simply allow users to manually configure the exceptions as they see fit. None of these solutions is perfect.

**Comment by vrslev** [comment]
If this certain issue is a "wontfix", I think it should be documented.

**Comment by Goldziher** [comment]
Fair point. Do you want to contribute this?

**Comment by vrslev** [comment]
Yes. Is "Middleware" page is the best one where to put this information?

**Comment by Goldziher** [comment]
sure

**Comment by vrslev** [comment]
Thanks for your effort to help with this issue and passion for the project, by the way! ❤️

**Comment by Goldziher** [comment]
Ok, i updated the docs - will be deployde shortly.
